File: src/main/webapp/resources/js/controllers/MidpController.js

'use strict';


/**
 * CarController
 * @constructor
 */
var MidpController = function($scope, $http) {

    $scope.isSampleArrayMaxValue = false;

    $scope.values = {
        input_a: '',
        input_b: 12,
        input_c: '',
        input_d: '',
        input_e: '',
    };


    $scope.samples_array = [];

            $scope.add = function () {
              $scope.samples_array.push({
                physical: "",
                drops: ""
              });

              if ($scope.samples_array.length == 10){
                  $scope.isSampleArrayMaxValue = true;
              }
            };




    //get data from server using data sended
    $scope.getMidp = function(data) {

        var i = 0;
        //just for testing
        for (i = 0; i < $scope.samples_array.length ; i++){
            $scope.values["physical_" + i] = $scope.samples_array[i].physical;
            $scope.values["drops_" + i] = $scope.samples_array[i].drops;

        }


        $http.get('api/v1/midp/', {params:$scope.values}).success(function(results){
            $scope.results = results;
            console.log(results);
        });

        $scope.values = {
                input_a: '',
                input_b: 12,
                input_c: '',
                input_d: '',
                input_e: "",
        };

        $scope.samples_array = [];

    };

    $scope.isThisDisabled=function(){

        var isAvailable = $scope.isArrayValid();


        if (isAvailable == false){
            return true;
        }

        var sr = "";

        for(var i in $scope.values)
        {
            if (i == "input_e"){
                if ($scope.values[i] === undefined){
                    return true;
                }
            }
            else if (!$scope.values[i])
                    return true;
        }


       return false;
   }

   $scope.isArrayValid = function(){

       var i = 0;
       var size = $scope.samples_array.length;
       var isAvailable = false;

       var sr = "";

       if (size >= 2 && size <= 10 && size%2==0){
           for (i = 0; i < size; i++) {


                if (($scope.samples_array[i].physical != '' || $scope.samples_array[i].physical !== sr) && $scope.samples_array[i].drops != ''

                    ){
                        if ($scope.samples_array[i].physical >= 0 && $scope.samples_array[i].physical <= 10 &&
                            $scope.samples_array[i].drops >= 15 && $scope.samples_array[i].drops <= 100){
                                isAvailable = true;
                        }
                        else{
                            return false;
                        }
                }
                else{
                    return false;
                }
            }
        }
        return isAvailable;


   }


};

File: src/main/java/pt/dei/springmvcangularjs/controller/MidPController.java
package pt.dei.springmvcangularjs.controller;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;

import pt.dei.insulin_calculator_server.ws_manager.WebServicesManager;
import pt.dei.springmvcangularjs.models.MidpModel;
import pt.dei.springmvcangularjs.models.MidsModel;
import pt.dei.springmvcangularjs.models.Response;

@Controller
@RequestMapping("/api/v1/midp")
public class MidPController {


    @Autowired
    private static WebServicesManager wsServicesManager;

    public MidPController() {
    	wsServicesManager = new WebServicesManager();
	}

    @RequestMapping("/")
    public @ResponseBody Response getMidp(@RequestParam Map<String,String> params){
    	
    	
    	MidsModel midsObject = new MidsModel(params.get("input_a"), params.get("input_b"), params.get("input_c"), params.get("input_d"), "-1000");
    	
       	List<Integer> physicalsamples= treatParams(params, "physical_");
    	List<Integer> bloodsamples = treatParams(params, "drops_");
    	
    	MidpModel midpObject = new MidpModel(convertString(params.get("input_e")), physicalsamples, bloodsamples);    	
    	
    	if (!midsObject.validate()){
    		return new Response();
    	}
    	
    	Response response = wsServicesManager.execute(midpObject, midsObject);
    	
    	return response;

    }

    
    private List<Integer> treatParams(Map<String,String> params,String type){
    	int size = (params.size() - 5) / 2; 
    	
    	List<Integer> list = new ArrayList<Integer>();
    	
    	for (int i = 0;i < size ;i++){
    		list.add(convertString(params.get(type + i)));
    	}
    	
    	return list;
    }
    
    private int convertString(String s){

		try {
			return Integer.parseInt(s);
		}
		catch (NumberFormatException e) {
			return -1;
		}
	}

    @RequestMapping("/layout")
    public String getMidsPartialPage() {
        return "midp/layout";
    }



}


